### ISSUE TITLE
Evaluation of Effect with magic-do optimization is inconsistent with other monadic code

### ISSUE THREAD

**Issue by natefaubion** [purescript/purescript#3913]
The magic-do optimization can cause arguments to the initial function call in a do block to be deferred where they otherwise would be strictly evaluated.

https://try.purescript.org/?gist=f2629a6024e8a957eaf9d03baccbc1d9

This gist only logs for Maybe, and not for Effect. If you remove the additional `pure unit` they both will log. This is a refactoring hazard.

**Comment by paf31** [comment]
I'm don't see how this is a bug - evaluation order isn't specified, but even if it was, evaluating something of type `Effect _` won't perform any effects unless you're using `unsafePerformEffect` like you are here. The whole point of `Effect` is to prevent evaluation order from running effectful code. Is there anything odd going on if you don't use `unsafePerformEffect`?

**Comment by natefaubion** [comment]
The `unsafePerformEffect` is just a convenient way to observe it. This is identical to the issue with function composition inlining reducing sharing, where a monomorphic optimization does not preserve de-facto assumptions about strict evaluation semantics. Arguments to functions are deferred where they otherwise would not be.

**Comment by hdgarrood** [comment]
We probably should specify evaluation order though, right? Especially since the language is strict?

**Comment by paf31** [comment]
So what would be a real-world example of this causing a problem?

**Comment by natefaubion** [comment]
Another way to state it is how can I reason about the desugaring of this do-block:
```
do
  ok (wat "hello")
  pure unit
```
It should be
```
ok (wat "hello") >>= \_ -> pure unit
```
Where the argument to `ok` (`wat "hello"`) is evaluated first.

The magic do optimizations turns this into the equivalent of
```
pure unit >>= \_ -> ok (wat "hello") >>= \_ -> pure unit
```
Where the call to `ok` and thus evaluation of it's arguments are completely deferred.

> So what would be a real-world example of this causing a problem?

My example was in a refactoring where code I had written was "optimal", and I removed the equivalent of `pure unit` at the end, and suddenly I had a major performance regression. It's the same core issue as eta-expansion with function composition.

**Comment by paf31** [comment]
Makes sense, thanks for clarifying. Then I think specifying that `do a <- x; y` desugars to something equivalent to `x >>= \a -> y` in every case would be useful. It might be tricky to implement that and keep the current optimization for `Effect` though, because if you pull the evaluation of `x` outside the `(function() { ... })`, then it's possible that it gets inlined again anyway. Hopefully the inliner doesn't do that, but I can't remember the details.

**Comment by natefaubion** [comment]
The optimization already keeps initial `let`s outside of the inserted `__do`, so that
```
example = do
  let arg = wat "hello"
  ok arg
  pure unit
```
Has the expected evaluation. I believe then that desired transformation would be to turn:
```
example = do
  ok (wat "hello")
  pure unit
```
into
```
example = do
  let initDoStatement = ok (wat "hello")
  initDoStatement
  pure unit
```
